### PR TITLE
Upgrade jest to v29

### PR DESCRIPTION
**Pull request by eps1lon**
Waiting for release of https://github.com/facebook/jest/pull/13328

## Summary

- unhandled errors from JSDOM are now proper errors instead of strings
- no more `abortSignal.reason = ...` hack in JSDOM environments 
- JSDOM environment by default now

## Test Plan

- [x] CI green

### CODE REVIEW

**Review by crypt096 — APPROVED** [review]

File: packages/react-dom/src/__tests__/ReactDOMEventPropagation-test.js [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
 'use strict';
 
-describe('ReactDOMEventListener', () => {
+describe('ReactDOMEventPropagation', () => {
```
Comment (eps1lon) [inline packages/react-dom/src/__tests__/ReactDOMEventPropagation-test.js:10]: Drive-by change. The suite name now matches the file name.

File: packages/react-dom/src/__tests__/ReactDOMEventPropagation-test.js [hunk 2]
```diff
@@ -16,12 +16,12 @@ describe('ReactDOMEventListener', () => {
...
     jest.resetModules();
-    React = require('react');
     jest.isolateModules(() => {
-      OuterReactDOM = require('react-dom');
+      InnerReactDOM = require('react-dom');
     });
     jest.isolateModules(() => {
-      InnerReactDOM = require('react-dom');
+      React = require('react');
+      OuterReactDOM = require('react-dom');
```
Comment (eps1lon) [inline packages/react-dom/src/__tests__/ReactDOMEventPropagation-test.js:24]: Fixes https://app.circleci.com/pipelines/github/facebook/react/21294/workflows/b23b2fbb-9d67-447c-b395-21591b60ebf1/jobs/411166.

Probably required due to https://github.com/facebook/jest/pull/10963. Though it's unclear why this test passed without this change using `scripts/jest/config.source-www.js` but failed using `scripts/jest/config.build.js`.

@SimenB Any idea if this is intended?

**Review by gaearon — APPROVED** [review]

File: packages/react-dom/src/__tests__/ReactDOMEventPropagation-test.js (hunk 2 above)
Comment (SimenB) [reply packages/react-dom/src/__tests__/ReactDOMEventPropagation-test.js:24]: Sounds odd that changing config impacts module loading. My only guess is https://github.com/facebook/react/blob/c1220ebdde506de91c8b9693b5cb67ac710c8c89/scripts/jest/config.build.js#L38-L46 makes it resolve to differing modules somehow?

Comment (eps1lon) [reply packages/react-dom/src/__tests__/ReactDOMEventPropagation-test.js:24]: The modules were all coming from the mapped paths. 
To be clear: It makes sense that the used `React` needs to be the same one that's used by `OuterReactDOM` and therefore needs to be in the same `isolatedModules` callback. What's odd is, that this isn't the case with `scripts/jest/config.source-www.js`.

So ideally we'd have two isolated copies of `react-dom` that somehow share the same `react` copy. This no longer seems possible with Jest 27

**Comment by eps1lon** [comment]
Some CI shards are failing with out-of-memory errors. Going to take a while identifying the memory leaks.

**Comment by gaearon** [comment]
Anything blocking this?

**Comment by eps1lon** [comment]
> Anything blocking this?

I noticed an issue when running `updateSnapshot` on `useId` tests (running `--updateSnashot` without https://github.com/facebook/react/pull/21575/commits/e44ea790c98e79804c40619ffd8a6b7ef3869f40). So I'd like to check if this is already broken on `main` or new with Jest 27.

**Comment by eps1lon** [comment]
@gaearon https://github.com/facebook/jest/issues/11561 is blocking this update. Jest without `--updateSnapshot` would be quite boring.

**Comment by SimenB** [comment]
Easiest to unblock is probably to float a patch (e.g. `patch-package`) just removing https://www.runpkg.com/?jest-snapshot@27.5.1/build/InlineSnapshots.js#222. The fix for the above issue will be in Jest 29, but going from 26 straight to 29 sounds like a lot 😀

**Comment by eps1lon** [comment]
@SimenB Currently stuck on 28. Seems like Jest now requires file extensions in the imports?
```
ENOENT: no such file or directory, open '/home/eps1lon/Development/forks/react/packages/react/src/ReactSharedInternals'

      146 | // the React package itself.
      147 | jest.mock('shared/ReactSharedInternals', () =>
    > 148 |   jest.requireActual('react/src/ReactSharedInternals')
          |          ^
      149 | );
      150 |
      151 | jest.mock('scheduler', () => jest.requireActual('scheduler/unstable_mock'));
```

**Comment by SimenB** [comment]
I'd guess it's somehow some weird `moduleNameMapper`: https://github.com/facebook/react/blob/3613284dceac38de4ef5a7c93c4897600ee6923b/scripts/jest/config.build.js#L57

Not 100% sure tho - I can take a look

**Comment by SimenB** [comment]
Ah, it's https://github.com/facebook/react/blob/3613284dceac38de4ef5a7c93c4897600ee6923b/packages/react/package.json#L22-L31

Should improve the error to specify it's due to no exports entry

**Comment by eps1lon** [comment]
> I'll fix in Jest 29, but if you wanna land v28 you'll need patch-package or something (or add the extension).

We need 29 for the JSX inline-snapshot fix so that's fine

**Comment by SimenB** [comment]
Release of fix for the resolver issue will come today 🙂

**Comment by SimenB** [comment]
https://github.com/facebook/jest/releases/tag/v29.0.3

**Comment by eps1lon** [comment]
@SimenB https://github.com/facebook/jest/pull/13328 seems to be the last blocker. Full test suite was green now locally with https://github.com/facebook/jest/pull/13328 applied.

File: packages/react-dom/src/__tests__/ReactDOMFizzServerBrowser-test.js [hunk 3]
```diff
@@ -272,22 +266,14 @@ describe('ReactDOMFizzServerBrowser', () => {
     } catch (error) {
       caughtError = error;
     }
-    expect(caughtError.message).toBe(
-      'The render was aborted by the server without a reason.',
-    );
-    expect(errors).toEqual([
-      'The render was aborted by the server without a reason.',
-    ]);
+    expect(caughtError.message).toBe('The operation was aborted.');
```
Comment (eps1lon) [inline packages/react-dom/src/__tests__/ReactDOMFizzServerBrowser-test.js:269]: Might be worth a follow-up. Currently Fizz thinks `controller.abort()` means that `signal.reason === undefined`. But the reason is only undefined if we didn't abort yet. If we did abort, reason will either be the value given in `abort` or a DOMException if reason is `undefined` (see https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/reason)

`controller.abort()` i.e. "no reason"
```bash
let controller = new AbortController()
undefined
controller.signal.reason
undefined
controller.abort()
undefined
controller.signal.reason
DOMException: signal is aborted without reason
    at <anonymous>:1:12
controller.signal.reason.message
'signal is aborted without reason'
```

`controller.abort('reason')`:
```bash
controller = new AbortController()
AbortController {signal: AbortSignal}
controller.signal.reason
undefined
controller.abort('the reason')
undefined
stats.ts:42          POST https://api.github.com/_private/browser/stats net::ERR_BLOCKED_BY_CLIENT
j @ stats.ts:42
requestIdleCallback (async)
i @ stats.ts:24
await in i (async)
g @ stats.ts:15
(anonymous) @ report-long-tasks.ts:6
controller.signal.reason
'the reason'
```

File: packages/react-dom/src/__tests__/ReactDOMHydrationDiff-test.js [hunk 4]
```diff
@@ -816,10 +820,10 @@ describe('ReactDOMServerHydration', () => {
         }
 
         expect(testMismatch(Mismatch)).toMatchInlineSnapshot(`
-            Array [
-              "Caught [The server did not finish this Suspense boundary: The server used \\"renderToString\\" which does not support Suspense. If you intended for this Suspense boundary to render the fallback content on the server consider throwing an Error somewhere within the Suspense boundary. If you intended to have the server wait for the suspended component please switch to \\"renderToPipeableStream\\" which supports Suspense on the server]",
-            ]
-          `);
+          Array [
+            "Caught [The server did not finish this Suspense boundary: The server used \\"renderToString\\" which does not support Suspense. If you intended for this Suspense boundary to render the fallback content on the server consider throwing an Error somewhere within the Suspense boundary. If you intended to have the server wait for the suspended component please switch to \\"renderToReadableStream\\" which supports Suspense on the server]",
```
Comment (eps1lon) [inline packages/react-dom/src/__tests__/ReactDOMHydrationDiff-test.js:824]: Looks like a fix to me. Tests now run in JSDOM by default so it suggests the browser API.

File: scripts/jest/config.base.js [hunk 5]
```diff
@@ -20,10 +20,20 @@ module.exports = {
...
   collectCoverageFrom: ['packages/**/*.js'],
-  timers: 'fake',
+  // TODO: Use the default `modern` implementation (includes `jest.useFakeTimers()`)
+  fakeTimers: {
+    enableGlobally: true,
+    legacyFakeTimers: true,
+  },
+  snapshotFormat: {
+    escapeString: true,
+    printBasicPrototype: true,
```
Comment (eps1lon) [inline scripts/jest/config.base.js:30]: Probably want to follow-up, use the new defaults and see if you like the snapshots better.

File: packages/scheduler/src/__tests__/SchedulerSetTimeout-test.js [hunk 6]
```diff
@@ -21,8 +21,9 @@ describe('SchedulerNoDOM', () => {
   // This is only meant to be used for testing purposes, like with jest's fake timer API.
   beforeEach(() => {
     jest.resetModules();
-    jest.useFakeTimers();
+    jest.useFakeTimers({legacyFakeTimers: true});
     delete global.setImmediate;
+    delete global.MessageChannel;
```
Comment (eps1lon) [inline packages/scheduler/src/__tests__/SchedulerSetTimeout-test.js:26]: `MessageChannel` is now defined in the Node.js environment but the test here specifically targets the `setTImeout` fallback.

File: packages/react-dom/src/__tests__/ReactDOMEventPropagation-test.js (hunk 2 above)
Comment (SimenB) [reply packages/react-dom/src/__tests__/ReactDOMEventPropagation-test.js:24]: Seems you've worked around this, but a solution might be a new API which lets you selectively evict specific modules from the cache? `jest.resetModule('react-dom')` or something?

Comment (eps1lon) [reply packages/react-dom/src/__tests__/ReactDOMEventPropagation-test.js:24]: I haven't thought about this for a while but if I remember correctly, I came to the conclusiong that the old code should never have worked to begin with. Not sure if it's worth revisiting since this testing pattern is very specific.

Comment (SimenB) [reply packages/react-dom/src/__tests__/ReactDOMEventPropagation-test.js:24]: That sounds very reasonable 👍 It does sound like a weird edge case, but happy to revisit it if proves to actually guard against some regression 🙂

File: packages/react-devtools-timeline/src/content-views/utils/__tests__/colors-test.js [hunk 7]
```diff
@@ -40,7 +40,8 @@ describe(dimmedColor, () => {
 describe(ColorGenerator, () => {
   describe(ColorGenerator.prototype.colorForID, () => {
     it('should generate a color for an ID', () => {
-      expect(new ColorGenerator().colorForID('123')).toMatchInlineSnapshot(`
+      const color = new ColorGenerator().colorForID('123');
+      expect(color).toMatchInlineSnapshot(`
```
Comment (eps1lon) [inline packages/react-devtools-timeline/src/content-views/utils/__tests__/colors-test.js:44]: Another example of https://github.com/facebook/jest/issues/6744

File: packages/use-sync-external-store/src/__tests__/useSyncExternalStoreShared-test.js [hunk 8]
```diff
@@ -882,8 +882,11 @@ describe('Shared useSyncExternalStore behavior (shim and built-in)', () => {
 
   describe('selector and isEqual error handling in extra', () => {
     let ErrorBoundary;
-    beforeAll(() => {
-      spyOnDev(console, 'warn');
+    beforeEach(() => {
+      // TODO There's a "Unsafe read of a mutable value during render." somewhere.
```
Comment (eps1lon) [inline packages/use-sync-external-store/src/__tests__/useSyncExternalStoreShared-test.js:886]: Found it: https://github.com/facebook/react/pull/25338#discussion_r982197631

File: packages/react-dom/src/__tests__/ReactDOMFizzStaticNode-test.js [hunk 9]
```diff
@@ -22,6 +20,20 @@ describe('ReactDOMFizzStaticNode', () => {
...
     if (__EXPERIMENTAL__) {
       ReactDOMFizzStatic = require('react-dom/static');
+      const AbortControllerPonyfill = require('abortcontroller-polyfill/dist/cjs-ponyfill')
+        .AbortController;
+      global.AbortController = class PatchedAbortController extends AbortControllerPonyfill {
+        abort(reason) {
+          // TODO: upstream this is a hack to work around lack of support for abortSignal.reason in the ponyfill
+          if (reason === undefined) {
+            this.signal.reason = new Error('This operation was aborted');
+            this.signal.reason.name = 'AbortError';
```
Comment (eps1lon) [inline packages/react-dom/src/__tests__/ReactDOMFizzStaticNode-test.js:30]: https://runkit.com/eps1lon/abortcontroler-abort-in-node

Comment (eps1lon) [reply packages/react-dom/src/__tests__/ReactDOMFizzStaticNode-test.js:30]: Upstream fix: https://github.com/mo/abortcontroller-polyfill/pull/65

**Comment by SimenB** [comment]
🟢 🥳

File: packages/react-dom/src/__tests__/ReactDOMFizzServer-test.js [hunk 10]
```diff
@@ -22,58 +24,74 @@ let useSyncExternalStoreWithSelector;
...
+
+function resetJSDOM(markup) {
+  // Test Environment
+  const virtualConsole = new JSDOMVirtualConsole();
+  virtualConsole.sendTo(console, {
+    omitJSDOMErrors: true,
+  });
+  virtualConsole.on('jsdomError', error => {
+    console.error(error);
+  });
```
Comment (gnoff) [inline packages/react-dom/src/__tests__/ReactDOMFizzServer-test.js:75]: @eps1lon what are these for? I'm not well versed in jsdom-isms but it's confusing you need to omit jsdom errors in the sendTo and then you rewrite them manually to the console. Does it always route to .log and you want .error instead?

Comment (eps1lon) [reply packages/react-dom/src/__tests__/ReactDOMFizzServer-test.js:75]: So if I understood these correctly, they're not actually omitting the errors. Just ensuring that they're forwarded in a way that is more "idiomatic". Without that, `jsdom` will issue `console.errors` with multiple arguments, but no placeholders. Thus triggering https://github.com/facebook/react/blob/3de926449681015c4f1eec1147baa03a3cb823f5/scripts/jest/matchers/toWarnDev.js#L236-L248

What we're doing here, is just recreating the JSDOM environment with the given HTML i.e. this is how `jest-environment-jsdom` already works: https://github.com/facebook/jest/blob/e21c5aba950f6019bbfde2f8233ac96d1fcaef42/packages/jest-environment-jsdom/src/index.ts#L41-L45

File: packages/react-dom/src/__tests__/ReactDOMFloat-test.js [hunk 11]
```diff
@@ -470,7 +484,6 @@ describe('ReactDOMFloat', () => {
           <head />
           <body>
             foo
-            <link rel="preload" as="style" href="foo" />
```
Comment (eps1lon) [inline packages/react-dom/src/__tests__/ReactDOMFloat-test.js:473]: According to the test, its actions should be run on the server. But as far as I can tell, there's only a client dispatcher (`ReactDOMClientDispatcher`).

@gnoff Do you have a e2e test for this scenario so that we can verify where the test environment might not be setup properly?

Jest has become a lot stricter with mixing environments and these sort of tests break due to strict environment separation. IMO this is ok behavior from Jest since it makes it more obvious where particular code is running.

**Comment by SimenB** [comment]
[React Native](https://github.com/facebook/react-native/pull/34724) and [Metro](https://github.com/facebook/metro/commit/69607ae715626f11eaecab1e0c05857c85f20662) have both upgraded to v29. Would be great if React could as well! 😀

**Comment by ymqy** [comment]
@eps1lon Why hasn't this pr been merged yet? I submitted a pr and all the ci tests have passed. It includes more optimization upgrades. https://github.com/facebook/react/pull/26041

**Comment by eps1lon** [comment]
Because it has merge conflicts and was not reviewed in its latest state
